### ISSUE TITLE
java.lang.NoClassDefFoundError: org/eclipse/jetty/client/api/Response$Listener

### ISSUE THREAD

**Issue by rsb5003** [dropwizard/dropwizard#1992]
When using dropwizard 1.0.7 with jetty-proxy 9.3.9.v20160517 I am seeing a NoClassDefFoundError when instantiating the ProxyServlet.Transparent class. 

```java
Exception in thread "main" java.lang.NoClassDefFoundError: org/eclipse/jetty/client/api/Response$Listener
	at com.comcast.ace.ProxyServerApplication.addProxyServlet(ProxyServerApplication.java:52)
	at com.comcast.ace.ProxyServerApplication.lambda$addProxyServlets$0(ProxyServerApplication.java:45)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.comcast.ace.ProxyServerApplication.addProxyServlets(ProxyServerApplication.java:45)
	at com.comcast.ace.ProxyServerApplication.run(ProxyServerApplication.java:40)
	at com.comcast.ace.ProxyServerApplication.run(ProxyServerApplication.java:17)
	at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:43)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:85)
	at io.dropwizard.cli.Cli.run(Cli.java:75)
	at io.dropwizard.Application.run(Application.java:79)
	at com.comcast.ace.ProxyServerApplication.main(ProxyServerApplication.java:20)
Caused by: java.lang.ClassNotFoundException: org.eclipse.jetty.client.api.Response$Listener
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 11 more
```

Upgrading to dropwizard 1.1 with jetty-proxy 9.4.2.v20170220 fixed the problem for me. 

The issue can be reproduced from the following repository: https://github.com/rsb5003/proxy-server-test.

**Comment by joschi** [comment]
@rsb5003 Try adding the missing [`jetty-client`](https://search.maven.org/#artifactdetails%7Corg.eclipse.jetty%7Cjetty-client%7C9.3.9.v20160517%7Cjar) dependency to your build.

**Comment by rsb5003** [comment]
Apologies for initially forgetting to add that when putting together this test project. But that doesn't resolve the issue. I've updated the project to include it for you to see. 

When debugging this problem, I tried several different variations of versions between dropwizard and jetty. In certain cases I'll get a different NoClassDefFoundError. For example if you upgrade to the jetty version to 9.4.2.v20170220, keeping dropwizard at 1.0.7, you'll get the below error. But the only solution I found to resolve all NoClassDefFoundErrors and be able to use the ProxyServlet.Transparent class is to upgrade dropwizard to 1.1

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/eclipse/jetty/client/ProtocolHandler
	at com.comcast.ace.ProxyServerApplication.addProxyServlet(ProxyServerApplication.java:52)
	at com.comcast.ace.ProxyServerApplication.lambda$addProxyServlets$0(ProxyServerApplication.java:45)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.comcast.ace.ProxyServerApplication.addProxyServlets(ProxyServerApplication.java:45)
	at com.comcast.ace.ProxyServerApplication.run(ProxyServerApplication.java:40)
	at com.comcast.ace.ProxyServerApplication.run(ProxyServerApplication.java:17)
	at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:43)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:85)
	at io.dropwizard.cli.Cli.run(Cli.java:75)
	at io.dropwizard.Application.run(Application.java:79)
	at com.comcast.ace.ProxyServerApplication.main(ProxyServerApplication.java:20)
Caused by: java.lang.ClassNotFoundException: org.eclipse.jetty.client.ProtocolHandler
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 11 more
```

**Comment by joschi** [comment]
> For example if you upgrade to the jetty version to 9.4.2.v20170220, keeping dropwizard at 1.0.7, you'll get the below error.

Yes, because Dropwizard 1.07 is **not using** Jetty 9.4.2.v20170220. You have to use a compatible dependency in your project, i. e. jetty-client 9.3.9.v20160517.

That's not a problem with Dropwizard, it's a problem with incompatible dependencies (and that jetty-proxy didn't seem to depend on jetty-client in version 9.3.9.v20160517).

**Comment by rsb5003** [comment]
So I guess my point in this ticket is that if you are using dropwizard 1.0.7, and the compatible jetty version 9.3.9.v20160517 functionality is broken and you can't use the ProxyServlet.Transparent class...

Not trying to be difficult. Just thought this would be a problem the dropwizard team would want to know about. If there is a solution to this problem, other than upgrading to dropwizard 1.1.0, I'd be interested in learning it. But in the meantime I'll plan on upgrading to 1.1.0 in order to solve this problem.

**Comment by joschi** [comment]
> So I guess my point in this ticket is that if you are using dropwizard 1.0.7, and the compatible jetty version 9.3.9.v20160517 functionality is broken and you can't use the ProxyServlet.Transparent class.

@rsb5003 Have you tried adding the jersey-client dependency in the correct version, as I suggested?

**Comment by joschi** [comment]
@rsb5003 
> I have added the jetty-client dependency as you suggested (your last post said 'jersey' which I don't think you meant)

Correct, that was a typo.

I think I've found out why it doesn't work in Dropwizard 1.0.x. The `jetty-client` artifact is defined in `dropwizard-bom` to be included in the `test` scope, which isn't available during runtime (or compilation of the `main` sources).

Try applying the following diff to your test project:

```diff
diff --git a/pom.xml b/pom.xml
index 38d9e8a..4be8c16 100644
--- a/pom.xml
+++ b/pom.xml
@@ -42,6 +42,7 @@
             <groupId>org.eclipse.jetty</groupId>
             <artifactId>jetty-client</artifactId>
             <version>${jetty.version}</version>
+            <scope>compile</scope>
         </dependency>
         <dependency>
             <groupId>org.eclipse.jetty</groupId>
```

**Comment by joschi** [comment]
Since using Dropwizard 1.1.x seems to work, we will not fix this nuisance in Dropwizard 1.0.x.

**Comment by rsb5003** [comment]
@joschi I updated my test project with the <scope>compile</scope> as you recommended. That solved the problem with being able to instantiate the ProxyServlet.Transparent class!

Thanks for you help in looking into this problem! I think my team will likely upgrade to dropwizard 1.1.0 anyway, however this is great to know there is a possible solution while still using dropwizard 1.0.x
